README.md
# Eventum SCM hook scripts

## CVS

Setup in your CVS server:

###  CVS 1.11:

 * `CVSROOT/loginfo`:

```
# process any message with eventum
ALL /path/to/eventum-cvs-hook.php -n cvs http://eventum.example.org/ $USER %{sVv}
```

###  CVS 1.12:

 * `CVSROOT/loginfo`:
```
# process any message with eventum
ALL /path/to/eventum-cvs-hook.php -n cvs http://eventum.example.org/ $USER "%p" %{sVv}
```
 * `CVSROOT/config`:
```
UseNewInfoFmtStrings=yes
```

## SVN

 * Setup in your svn server `hooks/post-commit`:

```sh
#!/bin/sh
REPO="$1"
REV="$2"
/path/to/eventum-svn-hook.php -n svn http://eventum.example.org/ "$REPO" "$REV"
```

## Git


 * Setup in your git repo `hooks/post-receive`:

```sh
#!/bin/sh
/path/to/eventum-git-hook.php -n git http://eventum.example.org/
```
## GitLab

GitLab is supported by Eventum itself, without scripts from this project.

Configure project in GitLab webhook to post to:
 - `https://eventum.example.org/scm_ping.php`

Recommended events are:
 - `push_events`: true
 - `tag_push_events`: true
 - `note_events`: true
 - `issues_events`: true
 - `merge_requests_events`: true

## Re-submit failed payload

In case Eventum server down, hooks create dump of post context which can be re-submitted to Eventum server.

```sh
/path/to/eventum-cvs-hook.php -n cvs http://eventum.example.org/ -l /tmp/eventum-cvs-hookuKrXLh
```

This is supported by CVS hook.


CHANGELOG.md
# Eventum SCM hook scripts

## [3.2.1] - 2020-10-20

- Fix Undefined index: program notice, #9

[3.2.1]: https://github.com/eventum/scm/compare/v3.2.0...v3.2.1

## [3.2.0] - 2019-08-20

- CVS: Add support for loading saved context, #7

[3.2.0]: https://github.com/eventum/scm/compare/v3.1.4...v3.2.0

## [3.1.4] - 2019-08-13

- include time in `save_environment`, #4

[3.1.4]: https://github.com/eventum/scm/compare/v3.1.3...v3.1.4

## [3.1.3] - 2017-10-24

- use --no-renames for git status extract, #3

[3.1.3]: https://github.com/eventum/scm/compare/v3.1.2...v3.1.3

## [3.1.2] - 2016-09-06

- git hook does not match multiline messages, #2

[3.1.2]: https://github.com/eventum/scm/compare/v3.1.1...v3.1.2

## [3.1.1] - 2016-09-06

- save execution environment in case of errors
- use 5s timeout for making requests
- automate .phar building using travis

[3.1.1]: https://github.com/eventum/scm/compare/v3.1.0...v3.1.1

## [3.1.0] - 2016-04-28

This version of hooks require Eventum 3.1.0.
From this version onwards eventum-scm is not released as part of main Eventum release.

- cvs/svn/git hooks rewritten to handle new payload for Eventum 3.1.0
- use JSON payload to post data to Eventum, [134fd35]
- lose config, use command line arguments to configure hooks
- use GL_REPO as git project name

[3.1.0]: https://github.com/eventum/scm/compare/v3.0.12...v3.1.0
[134fd35]: https://github.com/eventum/scm/commit/134fd35

## [3.0.12] - 2016-04-19

- cvs/svn/git: include commitid information on commits

[3.0.12]: https://github.com/eventum/scm/compare/v3.0.4...v3.0.12

## [3.0.4] - 2015-10-31

- no functional changes

[3.0.4]: https://github.com/eventum/scm/compare/v3.0.3...v3.0.4

## [3.0.3] - 2015-10-13

- git: default to `GL_REPO`, [6a5d58c]

[3.0.3]: https://github.com/eventum/scm/compare/v3.0.2...v3.0.3
[6a5d58c]: https://github.com/eventum/scm/commit/6a5d58c

## [3.0.2] - 2015-08-04

- fix cvs hook handling in old (1.11) cvs server, [2a83268]
- add makefile and box configuration to build standalone .phar [7927586]

[3.0.2]: https://github.com/eventum/scm/compare/v3.0.0-pre1...v3.0.2
[2a83268]: https://github.com/eventum/scm/commit/2a83268
[7927586]: https://github.com/eventum/scm/commit/7927586

## [3.0.0-pre1] - 2014-11-30

- moved out of Eventum sourcetree as separate git repo
- added git hook script [0e6c1ea], [LP#1078454]
- use json format to read scm ping responses
- restructure code to be more modular, move common code to helpers
- eventum now supports multiple scm repositories [d9f120b]

[3.0.0-pre1]: https://github.com/eventum/scm/compare/v2.4.0-pre1...v3.0.0-pre1
[LP#1078454]: https://bugs.launchpad.net/eventum/+bug/1078454
[0e6c1ea]: https://github.com/eventum/scm/commit/0e6c1ea
[d9f120b]: https://github.com/eventum/scm/commit/d9f120b

## [2.4.0-pre1] - 2014-10-04

- no functional changes

[2.4.0-pre1]: https://github.com/eventum/scm/compare/v2.3.6...v2.4.0-pre1

## [2.3.6] - 2014-01-28

- no changes

[2.3.6]: https://github.com/eventum/scm/compare/v2.3.5...v2.3.6

## [2.3.5] - 2014-01-24

- no changes

[2.3.5]: https://github.com/eventum/scm/compare/v2.3.4...v2.3.5

## [2.3.4] - 2013-11-28

- cvs: handle warnings better when `cvs add directory` is performed [44c9d4a]
- load local config for CVS/SVN integration from script dir [371973b]

[2.3.4]: https://github.com/eventum/scm/compare/v2.3.3...v2.3.4
[371973b]: https://github.com/eventum/scm/commit/371973b
[44c9d4a]: https://github.com/eventum/scm/commit/44c9d4a

## [2.3.3] - 2012-05-30

- cvs: support CVS 1.11 and 1.12 formats (autodetected if configured correctly) [62b7f9c]

[2.3.3]: https://github.com/eventum/scm/compare/v2.3.2...v2.3.3
[62b7f9c]: https://github.com/eventum/scm/commit/62b7f9c

## [2.3.2] - 2011-12-03

- use PHP's native `stream_get_contents()` instead of getInput function to read data from STDIN [2589912]

[2.3.2]: https://github.com/eventum/scm/compare/v2.3.1...v2.3.2
[2589912]: https://github.com/eventum/scm/commit/2589912

## [2.3.1] - 2011-02-10

- no changes

[2.3.1]: https://github.com/eventum/scm/compare/v2.3...v2.3.1

## [2.3] - 2010-07-28

- cvs: handle importing sources and adding new directories nicely [930e438]
- cvs: commit hook prints remote response [a83f486]

[2.3]: https://github.com/eventum/scm/compare/v2.2...v2.3
[930e438]: https://github.com/eventum/scm/commit/930e438
[a83f486]: https://github.com/eventum/scm/commit/a83f486

## [2.2] - 2009-01-14

- no functional changes

[2.2]: https://github.com/eventum/scm/compare/v2.1.1...v2.2

## [2.1.1] - 2008-01-09

- cvs: handle errors better from HTTP response [8e10cfe]

[2.1.1]: https://github.com/eventum/scm/compare/v2.1...v2.1.1
[8e10cfe]: https://github.com/eventum/scm/commit/8e10cfe

## [2.1] - 2007-11-20

- no functional changes

[2.1]: https://github.com/eventum/scm/compare/v2.0.1...v2.1

## [2.0.1] - 2007-04-17

- no functional changes

[2.0.1]: https://github.com/eventum/scm/compare/v2.0.0...v2.0.1

## [2.0.0] - 2007-04-12

- fix issue id detection in scm commit message [a567490]
- new config format, use `$eventum_url` variable, allows `https://` urls [b6bf3b1]
- drop use legacy `$HTTP_SERVER_VARS`
- initial svn hook script [698176a]
- allow SCM commit messages contain multiple issue IDs [e89814b]
- properly encode post data containing `+` (use `rawurlencode` instead of `base64_encode`) [ac8b3ee]
- report errors from from Eventum server [60304fb]

[2.0.0]: https://github.com/eventum/scm/compare/v1.7.1...v2.0.0
[60304fb]: https://github.com/eventum/scm/commit/60304fb
[698176a]: https://github.com/eventum/scm/commit/698176a
[a567490]: https://github.com/eventum/scm/commit/a567490
[ac8b3ee]: https://github.com/eventum/scm/commit/ac8b3ee
[b6bf3b1]: https://github.com/eventum/scm/commit/b6bf3b1
[e89814b]: https://github.com/eventum/scm/commit/e89814b


helpers.php
<?php

/*
 * This file is part of the Eventum (Issue Tracking System) package.
 *
 * @copyright (c) Eventum Team
 * @license GNU General Public License, version 2 or later (GPL-2+)
 *
 * For the full copyright and license information,
 * please see the COPYING and AUTHORS files
 * that were distributed with this source code.
 */

/**
 * Execute command
 *
 * @param string $command
 * @return array command output each line as array element
 * @throw RuntimeException throw exception if exits with non-zero
 */
function execx($command)
{
    exec($command, $output, $rc);
    if ($rc) {
        throw new RuntimeException("$command exited with $rc");
    }

    return $output;
}

/**
 * Execute command, returning first line from it
 *
 * @param string $command
 * @return string
 */
function execl($command)
{
    $output = execx($command);

    return current($output);
}

/**
 * Submit SCM commit data to Eventum.
 *
 * @param array $params
 */
function scm_ping($params)
{
    global $PROGRAM, $eventum_url;

    $ping_url = $eventum_url . 'scm_ping.php?scm=' . $params['scm'];
    $status = json_post($ping_url, $params, 1);

    if ($status['code']) {
        throw new RuntimeException($status['message'], $status['code']);
    }

    $message = trim($status['message']);
    if (!$message) {
        return;
    }

    // prefix response with our name
    foreach (explode("\n", $message) as $line) {
        echo "$PROGRAM: $line\n";
    }
}

/**
 * Extract dir and file name from abspath
 *
 * @param string $abspath
 * @return array file dirname and basename
 */
function fileparts($abspath)
{
    // special for "dirname/" case, pathinfo would set dir to '.' and filename to 'dirname'
    $length = strlen($abspath);
    if ($abspath[$length - 1] === '/') {
        return array(rtrim($abspath, '/'), '');
    }

    $fi = pathinfo($abspath);

    return array($fi['dirname'], $fi['basename']);
}

/**
 * parse the commit message and get all issue numbers we can find
 *
 * @param string $commit_msg
 * @param string|null $eventum_url
 * @return array
 */
function match_issues($commit_msg, $eventum_url = null)
{
    $url_quoted = rtrim($eventum_url ? preg_quote($eventum_url, '/') : '', '/');

    preg_match_all("/
        (?:
            # match issue xxx and bug xxx
            (?i:issue|bug)\s*:?\s*#?
            |
            # match url
            {$url_quoted}/view\.php\?id=
        )

        (?P<issue_id>\d+)
    /x", $commit_msg, $matches);

    if (count($matches['issue_id']) > 0) {
        return $matches['issue_id'];
    }

    return null;
}

/**
 * Fetch $url, return response and optionally unparsed headers array.
 *
 * @param string $url URL to request
 * @param array $params QueryString parameters to URL
 * @param bool $headers = false
 * @return mixed
 * @return array|string
 * @author Elan Ruusamäe <glen@delfi.ee>
 */
function wget($url, $params, $headers = true)
{
    $url .= '?' . http_build_query($params, null, '&');

    // see if we can fopen
    $flag = ini_get('allow_url_fopen');
    if (!$flag) {
        throw new RuntimeException('allow_url_fopen is disabled');
    }

    // see if https is supported
    $scheme = parse_url($url, PHP_URL_SCHEME);
    if (!in_array($scheme, stream_get_wrappers(), true)) {
        throw new RuntimeException("$scheme:// scheme not supported. Load openssl php extension?");
    }

    $fp = @fopen($url, 'r');
    if (!$fp) {
        $error = error_get_last();
        throw new RuntimeException($error['message']);
    }

    if ($headers) {
        $meta = stream_get_meta_data($fp);
    }

    $data = '';
    while (!feof($fp)) {
        $data .= fread($fp, 4096);
    }
    fclose($fp);

    if ($headers) {
        return array($meta['wrapper_data'], $data);
    }

    return $data;
}

/**
 * POST json encoded data to $url
 *
 * @param string $url
 * @param array $data
 * @param bool $assoc
 * @return array|stdClass result with extra 'meta' key
 * @author Elan Ruusamäe <glen@delfi.ee>
 */
function json_post($url, $data, $assoc = false)
{
    // see if schema in url is supported
    $scheme = parse_url($url, PHP_URL_SCHEME);
    if (!in_array($scheme, stream_get_wrappers())) {
        throw new RuntimeException("$scheme:// scheme not supported. Load openssl php extension?");
    }

    $body = json_encode($data);
    $headers = array(
        'Expect: ',
        'Content-Type: application/json',
        'Accept: application/json',
        'Content-Length: ' . strlen($body),
    );

    $options = array(
        'method' => 'POST',
        'content' => $body,
        'header' => implode("\r\n", $headers),
        'timeout' => 5.0,
    );
    $options = array(
        // this needs to be 'http', regardless if we post to https://
        'http' => $options,
    );

    $context = stream_context_create($options);

    $stream = @fopen($url, 'r', false, $context);
    if (!$stream) {
        $error = error_get_last();
        throw new RuntimeException($error['message']);
    }

    $result = stream_get_contents($stream);
    $meta = stream_get_meta_data($stream);
    fclose($stream);

    $response = json_decode($result, $assoc);
    if (!$response) {
        throw new InvalidArgumentException("Unable to decode: $result");
    }
    if ($assoc) {
        $response['meta'] = $meta;
        $response['raw'] = $result;
    } else {
        $response->raw = $result;
        $response->meta = $meta;
    }

    return $response;
}

/**
 * Sane getopt() ajusted from this post:
 * http://php.net/getopt#100573
 */
function _getopt($parameters)
{
    global $argv, $argc;

    $options = getopt($parameters);
    $pruneargv = array();
    foreach ($options as $option => $value) {
        foreach ($argv as $key => $chunk) {
            $regex = '/^' . (isset($option[1]) ? '--' : '-') . $option . '/';
            if (($chunk == $value && $argv[$key - 1][0] == '-') || preg_match($regex, $chunk)) {
                $pruneargv[] = $key;
            }
        }
    }
    while ($key = array_pop($pruneargv)) {
        unset($argv[$key]);
    }

    // renumber $argv to be continuous
    $argv = array_values($argv);
    // reset $argc to be correct
    $argc = count($argv);

    return $options;
}

/**
 * Static version to get STDIN more than once even for older PHP engines
 */
function getInput()
{
    static $stdin;
    if ($stdin === null) {
        $stdin = stream_get_contents(STDIN);
    }

    return $stdin;
}

/**
 * Retrieve environment variables
 * As $_ENV is not reliable (variables_order may not contain E), we use phpinfo() call
 */
function get_all_env()
{
    if (PHP_VERSION_ID >= 70100) {
        return getenv();
    }

    ob_start();
    phpinfo(INFO_ENVIRONMENT);
    $buffer = ob_get_clean();

    # parse output like: "CVS_PID => 27518"
    preg_match_all('/^(?P<name>[^=]+)\s=>\s+/m', $buffer, $m);

    # we use getenv() to get "raw" value of env
    $env = array();
    foreach ($m['name'] as $key) {
        $value = getenv($key);
        if ($value !== false) {
            $env[$key] = $value;
        }
    }

    return $env;
}

/**
 * Create execution environment context
 *
 * @param array $argv
 * @return array
 */
function create_context(array $argv)
{
    return array(
        'time' => microtime(true),
        'php_version' => PHP_VERSION,
        'argv' => $argv,
        'cwd' => getcwd(),
        'stdin' => getInput(),
        'env' => get_all_env(),
    );
}

/**
 * Store execution environment details to temp file so the failed command could be repeated
 *
 * @param array $context
 * @return bool|string
 */
function store_environment(array $context)
{
    global $PROGRAM;

    $tmpfile = tempnam(sys_get_temp_dir(), $PROGRAM);
    file_put_contents($tmpfile, serialize($context));

    return $tmpfile;
}

/**
 * @param array $argv
 * @return string
 */
function save_environment($argv)
{
    return store_environment(create_context($argv));
}

/**
 * Load context from file, and adjust environment accordingly.
 *
 * @param string $dump_file
 * @return array
 */
function load_context($dump_file)
{
    $contents = file_get_contents($dump_file);
    if ($contents === false) {
        throw new RuntimeException("Unable to load $dump_file");
    }

    /** @noinspection UnserializeExploitsInspection */
    $context = unserialize($contents);

    // backward compatible
    if (!isset($context['argv']) && isset($context['command'])) {
        $context['argv'] = $context['command'];
    }

    // backward compatible
    if (!isset($context['program'])) {
        $context['program'] = basename($context['argv'][0], '.php');
    }

    return $context;
}


eventum-git-hook.php
#!/usr/bin/env php
<?php

/*
 * This file is part of the Eventum (Issue Tracking System) package.
 *
 * @copyright (c) Eventum Team
 * @license GNU General Public License, version 2 or later (GPL-2+)
 *
 * For the full copyright and license information,
 * please see the COPYING and AUTHORS files
 * that were distributed with this source code.
 */

require_once __DIR__ . '/helpers.php';

$original_argv = $argv;
$default_options = array(
    // scm name
    'n' => 'git',
    // project name: default to $GL_REPO
    'p' => getenv('GL_REPO'),
);

$options = _getopt('n:p:') + $default_options;

$PROGRAM = basename(realpath(array_shift($argv)), '.php');
$eventum_url = array_shift($argv);
$scm_name = $options['n'];
$scm_project = $options['p'];

$reflist = git_receive_refs();
process_push($reflist);
exit(0);

function process_push($reflist)
{
    $nullsha1 = '0000000000000000000000000000000000000000';
    $emptysha1 = '4b825dc642cb6eb9a060e54bf8d69288fbee4904';

    // process each branch push
    foreach ($reflist as $refs) {
        list($old, $new, $refname) = $refs;
        if ($new === $nullsha1) {
            // remote branch is deleted. nothing to do
            continue;
        }

        if ($old === $nullsha1) {
            // remote branch is created. use emptysha1 instead
            $old = $emptysha1;
        }

        $revlist = git_rev_list($old, $new, '--no-merges --author-date-order --reverse');
        foreach ($revlist as $rev) {
            git_scm_ping($rev, $refname);
        }
    }
}

/**
 * Submit Git data to Eventum
 *
 * @param string $rev
 */
function git_scm_ping($rev, $refname)
{
    $commit_msg = git_commit_msg($rev);
    $issues = match_issues($commit_msg);
    if (!$issues) {
        return;
    }

    global $PROGRAM, $scm_name, $scm_project;
    $author_email = git_commit_author_email($rev);
    $author_name = git_commit_author_name($rev);
    $commit_date = git_commit_author_date($rev);
    $files = git_commit_files($rev);
    $branch = git_branch_name($refname);

    $params = array(
        'scm' => 'git',
        'scm_name' => $scm_name,
        'project' => $scm_project,
        'author_email' => $author_email,
        'author_name' => $author_name,
        'commit_date' => $commit_date,
        'branch' => $branch,
        'commit_msg' => $commit_msg,
        'issue' => $issues,
        'files' => $files,
        'commitid' => $rev,
    );

    try {
        scm_ping($params);
    } catch (Exception $e) {
        global $original_argv;
        error_log("ERROR[$PROGRAM]: " . $e->getMessage());
        error_log('Debug saved to: ' . save_environment($original_argv));
        exit(1);
    }
}

/*
 * A post-receive hook gets its arguments from STDIN,
 * in the form <oldrev> <newrev> <refname>.
 * Since these arguments are coming from stdin, not from a command line argument,
 * you need to use read instead of $1 $2 $3.
 *
 * The post-receive hook can receive multiple branches at once (for example if someone does a git push --all),
 * so we also need to wrap the read in a while loop.
 */
function git_receive_refs()
{
    $result = array();
    $input = getInput();
    foreach (explode(PHP_EOL, rtrim($input, PHP_EOL)) as $line) {
        $result[] = explode(' ', $line);
    }

    return $result;
}

/**
 * Get files affected from $rev
 *
 * @param string $rev
 * @return array
 */
function git_commit_files($rev)
{
    $file_info = execx("git show --pretty=format: --no-renames --name-status $rev");

    // man git-show --diff-filter
    $map = array(
        'A' => 'added',
        'D' => 'removed',
        'M' => 'modified',
    );
    $files = array_fill_keys(array_values($map), array());
    foreach ($file_info as $line) {
        list($status, $filename) = explode("\t", $line, 2);

        if (isset($map[$status])) {
            $change_type = $map[$status];
        } else {
            error_log("git-show: Unknown type: $line");
            $change_type = 'unknown';
        }

        $files[$change_type][] = $filename;
    }

    return $files;
}

/**
 * @param string $old revision start
 * @param string $new revision end
 * @param string $options
 * @return array
 */
function git_rev_list($old, $new, $options = '')
{
    return execx("git rev-list $old..$new $options");
}

/**
 * @param string $rev
 * @return string
 */
function git_commit_author_email($rev)
{
    return git_format($rev, '%ae');
}

/**
 * @param string $rev
 * @return string
 */
function git_commit_author_name($rev)
{
    return git_format($rev, '%an');
}

/**
 * @param string $rev
 * @return string
 */
function git_commit_author_date($rev)
{
    return git_format($rev, '%at');
}

/**
 * @param string $rev
 * @return string
 */
function git_commit_msg($rev)
{
    return implode("\n", execx("git log --format=%B -n1 $rev"));
}

/**
 * Pretty format $rev with $format. It should be one line out output
 *
 * @param string $rev
 * @return string
 */
function git_format($rev, $format)
{
    return execl("git log --format=$format -n1 $rev");
}

/**
 * get git branch name for the refname
 *
 * @param string $refname
 * @return string
 */
function git_branch_name($refname)
{
    return execl("git rev-parse --symbolic --abbrev-ref $refname");
}


eventum-svn-hook.php
#!/usr/bin/env php
<?php

/*
 * This file is part of the Eventum (Issue Tracking System) package.
 *
 * @copyright (c) Eventum Team
 * @license GNU General Public License, version 2 or later (GPL-2+)
 *
 * For the full copyright and license information,
 * please see the COPYING and AUTHORS files
 * that were distributed with this source code.
 */

require_once __DIR__ . '/helpers.php';

$original_argv = $argv;
$default_options = array(
    'n' => 'svn',
);
$options = _getopt('n:') + $default_options;

$PROGRAM = basename(realpath(array_shift($argv)), '.php');
$eventum_url = array_shift($argv);
$scm_name = $options['n'];

try {
    main($scm_name, $argv);
} catch (Exception $e) {
    error_log("ERROR[$PROGRAM]: " . $e->getMessage());
    error_log('Debug saved to: ' . save_environment($original_argv));
    exit(1);
}
exit(0);

function main($scm_name, $argv)
{
    if (count($argv) !== 2) {
        throw new InvalidArgumentException('Invalid arguments');
    }

    list($repos, $rev) = $argv;

    global $svnlook;
    if (!isset($svnlook)) {
        $svnlook = '/usr/bin/svnlook';
    }

    if (!is_executable($svnlook)) {
        throw new BadFunctionCallException('svnlook is not executable, edit $svnlook');
    }

    $results = svnlook('info', $repos, $rev);
    list($username, $date, $commit_msg) = svn_commit_info($results);

    // parse the commit message and get all issue numbers we can find
    $issues = match_issues($commit_msg);
    if (!$issues) {
        return;
    }

    $files = svn_commit_files($repos, $rev);

    $params = array(
        'scm' => 'svn',
        'scm_name' => $scm_name,
        'username' => $username,
        'commit_msg' => $commit_msg,
        'issue' => $issues,
        'files' => $files,
        'commitid' => $rev,
    );

    scm_ping($params);
}

/**
 * Process username, date and commit message from svnlook output
 *
 * @param array $results
 * @return array
 */
function svn_commit_info($results)
{
    // get commit date and username and commit message
    $username = array_shift($results);
    $date = array_shift($results);

    // ignore commit message length value
    array_shift($results);

    // get the full commit message
    $commit_msg = implode("\n", $results);

    return array($username, $date, $commit_msg);
}

/**
 * Get files affected from $rev
 *
 * @param string $rev
 * @return array
 */
function svn_commit_files($repo, $rev)
{
    // create array with predefined keys
    $files = array(
        'added',
        'removed',
        'modified',
    );
    $files = array_fill_keys($files, array());

    $changes = svnlook('changed', $repo, $rev);
    foreach ($changes as $change) {
        // http://svnbook.red-bean.com/en/1.7/svn.ref.svnlook.c.changed.html
        // flags:
        // - 'A ' Item added to repository
        // - 'D ' Item deleted from repository
        // - 'U ' File contents changed
        // - '_U' Properties of item changed; note the leading underscore
        // - 'UU' File contents and properties changed
        list($change_info, $filename) = preg_split('/\s+/', $change, 2);
        $flags = preg_split('//', $change_info, -1, PREG_SPLIT_NO_EMPTY);

        if (in_array('A', $flags, true)) {
            $change_type = 'added';
        } elseif (in_array('D', $flags, true)) {
            $change_type = 'removed';
        } else {
            $change_type = 'modified';
        }

        $files[$change_type][] = $filename;
    }

    return $files;
}

/**
 * Execute svnlook command on $repo for $revision
 *
 * @param string $command
 * @param string $repo
 * @param int $revision
 * @return array
 */
function svnlook($command, $repo, $revision)
{
    global $svnlook;

    return execx("$svnlook $command $repo -r $revision");
}


eventum-cvs-hook.php
#!/usr/bin/env php
<?php

/*
 * This file is part of the Eventum (Issue Tracking System) package.
 *
 * @copyright (c) Eventum Team
 * @license GNU General Public License, version 2 or later (GPL-2+)
 *
 * For the full copyright and license information,
 * please see the COPYING and AUTHORS files
 * that were distributed with this source code.
 */

require_once __DIR__ . '/helpers.php';

$default_options = array(
    'n' => 'cvs',
);
$options = _getopt('l:n:') + $default_options;

if (isset($options['l'])) {
    $context = load_cvs_context($options['l']);
} else {
    $context = create_context($argv);
}

$context['arguments'] = array_slice($context['argv'], 2);
$context['scm_name'] = $options['n'];

// these need to be global. for now
$eventum_url = $context['argv'][1];
$PROGRAM = basename($context['argv'][0], '.php');

try {
    main($context);
} catch (Exception $e) {
    error_log("ERROR[$PROGRAM]: " . $e->getMessage());
    error_log('Debug saved to: ' . store_environment($context));
    exit(1);
}
exit(0);

function main($context)
{
    $commit_msg = cvs_commit_msg($context['stdin']);

    // parse the commit message and get all issue numbers we can find
    $issues = match_issues($commit_msg, $context['eventum_url']);
    if (!$issues) {
        return;
    }

    list($username, $modified_files) = cvs_commit_info($context['arguments']);
    list($commitid, $files) = cvs_get_files($modified_files);

    $params = array(
        'scm' => 'cvs',
        'commit_date' => $context['time'],
        'scm_name' => $context['scm_name'],
        'username' => $username,
        'commit_msg' => $commit_msg,
        'issue' => $issues,
        'files' => $files,
        'commitid' => $commitid,
    );

    scm_ping($params);
}

/**
 * create files payload
 *
 * @param array $modified_files
 * @return array
 */
function cvs_get_files($modified_files)
{
    // create array with predefined keys
    $files = array(
        'added',
        'removed',
        'modified',
    );
    $files = array_fill_keys($files, array());

    $commitid = array();
    foreach ($modified_files as $file) {
        $filename = $file['filename'];

        if (!$file['old_revision']) {
            $change_type = 'added';
        } elseif (!$file['new_revision']) {
            $change_type = 'removed';
        } else {
            $change_type = 'modified';
        }

        $files['versions'][$filename] = array($file['old_revision'], $file['new_revision']);
        $files[$change_type][] = $filename;
        $commitid[] = $file['commitid'];
    }

    // flatten commitid
    $commitid = array_unique($commitid);
    if (count($commitid) > 1) {
        throw new InvalidArgumentException('Commit Id should be unique');
    }
    $commitid = current($commitid);

    return array($commitid, $files);
}

/**
 * @param array $argv
 * @return array of username, modified files
 */
function cvs_commit_info($argv)
{
    // user who is committing these changes
    $username = array_shift($argv);

    if (count($argv) === 1) {
        $modified_files = cvs_parse_info_1_11($argv);
    } else {
        $modified_files = cvs_parse_info_1_12($argv);
    }

    // parse told to skip (for example adding new dir)
    if (!$modified_files) {
        exit(0);
    }

    return array($username, $modified_files);
}

/**
 * assume the old way ("PATH {FILE,rev1,rev2 }+")
 * CVSROOT/loginfo: ALL eventum-cvs-hook $USER %{sVv}
 *
 * @return array
 */
function cvs_parse_info_1_11($argv)
{
    $args = explode(' ', array_shift($argv));

    // save what the name of the module is
    $cvs_module = array_shift($args);

    // skip if we're importing or adding new dirrectory
    $msg = implode(' ', array_slice($args, -3));
    if (in_array($msg, array('- Imported sources', '- New directory'))) {
        return null;
    }

    // now parse the list of modified files
    $modified_files = array();
    foreach ($args as $file_info) {
        list($filename, $old_revision, $new_revision) = explode(',', $file_info);
        $modified_files[] = array(
            'filename' => "$cvs_module/$filename",
            'old_revision' => cvs_filter_none($old_revision),
            'new_revision' => cvs_filter_none($new_revision),
        );
    }

    return $modified_files;
}

/**
 * assume the new way ("PATH" {"FILE" "rev1" "rev2"}+)
 * CVSROOT/loginfo: ALL eventum-cvs-hook $USER "%p" %{sVv}
 *
 * @return array
 */
function cvs_parse_info_1_12($args)
{
    // save what the name of the module is
    $cvs_module = array_shift($args);

    // skip if we're importing or adding new directory
    // TODO: checked old way with CVS 1.11, but not checked the new way
    $msg = implode(' ', array_slice($args, -3));
    if (in_array($msg, array('- Imported sources', '- New directory'))) {
        return null;
    }

    // now parse the list of modified files
    $modified_files = array();
    $use_rcs = !file_exists('CVS/Entries');
    while ($file_info = array_splice($args, 0, 3)) {
        list($filename, $old_revision, $new_revision) = $file_info;
        $old_revision = cvs_filter_none($old_revision);
        $new_revision = cvs_filter_none($new_revision);
        $modified_files[] = array(
            'filename' => "$cvs_module/$filename",
            'old_revision' => $old_revision,
            'new_revision' => $new_revision,
            'commitid' => $use_rcs ? rcs_commitid($filename, $old_revision, $new_revision) : cvs_commitid($filename),
        );
    }

    return $modified_files;
}

/**
 * @param string $pattern
 * @param array $input
 * @return string|null
 * @internal
 */
function cvs_match_commitid($pattern, array $input)
{
    // find line matching 'Commit Identifier'
    $lines = preg_grep($pattern, $input);
    if (!$lines) {
        return null;
    }

    // match commit id
    if (!preg_match($pattern, current($lines), $m)) {
        return null;
    }

    return $m['commitid'];
}

/**
 * Extract 'commitid' from file using rcs
 *
 * @param string $filename
 * @param string|null $r1
 * @param string|null $r2
 * @return string
 */
function rcs_commitid($filename, $r1, $r2)
{
    if ($r2) {
        // added or modified file
        $result = execx('rcs log -r' . escapeshellarg($r2) . ' ' . escapeshellarg($filename));
    } else {
        // deleted file
        // NOTE: as revision not given, this may not work if file is added again meanwhile
        $result = execx('rcs log ' . escapeshellarg("Attic/$filename"));
    }

    // date: 2019/08/13 09:04:57;  author: glen;  state: Exp;  lines: +2 -3; commitid: uoHAXoFNyk8CxQyB
    $pattern = '/^date:.+commitid:\s+(?P<commitid>\S+)/sm';

    return cvs_match_commitid($pattern, $result);
}

/**
 * Extract 'commitid' from file, Requires CVS 1.12+
 *
 * @param string $filename
 * @return string
 */
function cvs_commitid($filename)
{
    $result = execx('cvs -Qn status ' . escapeshellarg($filename));

    $pattern = '/Commit Identifier:\s+(?P<commitid>\S+)/';

    return cvs_match_commitid($pattern, $result);
}

/**
 * filter out NONE revision
 *
 * @param string $rev
 */
function cvs_filter_none($rev)
{
    if ($rev !== 'NONE') {
        return $rev;
    }

    return null;
}

/**
 * Obtain CVS commit message
 *
 * @return string
 */
function cvs_commit_msg($input)
{
    // get the full commit message
    $commit_msg = rtrim(substr($input, strpos($input, 'Log Message:') + strlen('Log Message:') + 1));

    return $commit_msg;
}

/**
 * @param string $dump_file
 * @return array
 */
function load_cvs_context($dump_file)
{
    $context = load_context($dump_file);

    // extract real cwd to use
    $root = preg_quote($context['env']['CVSROOT'], '/');
    if (preg_match("/^Update of (?P<cwd>$root.+)$/m", $context['stdin'], $m)) {
        $context['original_cwd'] = $context['cwd'];
        $context['cwd'] = $m['cwd'];
    }

    if (chdir($context['cwd']) === false) {
        throw new RuntimeException("Unable to change directory to: {$context['cwd']}");
    }

    return $context;
}
